Cypress tests for masthead.

ssilvert opened the pull request:
## Motivation
Add masthead tests, fixing https://github.com/keycloak/keycloak-admin-ui/issues/37

## Verification Steps
Run the tests.

## Checklist:

- [X] Code has been tested locally by PR requester
- [NA] User-visible strings are using the react-i18next framework (useTranslation)
- [NA] Help has been implemented
- [NA] axe report has been run and resulting a11y issues have been resolved
- [NA] Unit tests have been created/updated
- [X] Formatting has been performed via prettier/eslint
- [X] Type checking has been performed via 'yarn check-types'

ssilvert commented:
Cypress tests were running locally but failing here because it brings up the old account console.  Apparently, we are running these tests against an old version of Keycloak.  That was really easy to figure out based on the mp4 output from Cypress.

I'm really, really loving Cypress!

ssilvert commented:
@Aboullos @edewit I've tried to upgrade the version of Keycloak we are using by bumping the version in start.js.  However, that produced a URL that doesn't exist.  Changing it to the correct download URL didn't seem to help.  Since I don't really know how all that stuff works I'm probably just making a mess of things.  Can you guys help?

Aboullos commented:
I am getting an error when launching the ./start.js.
The first Cypress test uses the old console to create the admin user and add the new admin console as a client.

Aboullos commented on tests/cypress/integration/masthead_test.spec.js:
```diff
@@ -0,0 +1,70 @@
+
+const goToAcctMgtTest = (userDropdown) => {
+  it("opens manage account and returns to admin console", () => {
+    cy.get(userDropdown).click();
```
```suggestion
    cy.get(userDropdown).click();
    cy.get("#manage-account").click();
```

It is better to use PageObjects to reuse code, this two lines can be replaced by
```
    headerPage.goToManageAccount();
```

Aboullos commented on tests/cypress/integration/masthead_test.spec.js:
```diff
@@ -0,0 +1,70 @@
+
+const goToAcctMgtTest = (userDropdown) => {
+  it("opens manage account and returns to admin console", () => {
+    cy.get(userDropdown).click();
+    cy.get("#manage-account").click();
```
It is better to use PageObjects to reuse code, this two lines can be replaced by
```suggestion
    headerPage.goToManageAccount();
```

Aboullos commented on tests/cypress/integration/masthead_test.spec.js:
```diff
@@ -0,0 +1,70 @@
+  it("opens manage account and returns to admin console", () => {
+    cy.get(userDropdown).click();
+    cy.get("#manage-account").click();
+    cy.contains("Welcome to Keycloak Account Management");
+    cy.get("#landingReferrerLink").click({ force: true });
```
This can be set in a PageObject like `AccountConsolePage` as two methods `checkIsAccountConsole` and `goBackToAdminConsole`

Aboullos commented on tests/cypress/integration/masthead_test.spec.js:
```diff
@@ -0,0 +1,70 @@
+    cy.get("#manage-account").click();
+    cy.contains("Welcome to Keycloak Account Management");
+    cy.get("#landingReferrerLink").click({ force: true });
+    cy.get(userDropdown);
```
This one can be removed

edewit commented on start.js:
```diff
@@ -1,53 +1,55 @@
-      decompressKeycloak().then(() => run());
-    });
-  });
+  http.get(
```
The reason it fails is because of a 302 from github so we should follow that redirect:

```typescript
const request = (url, file) => {
  http.get(url, (response) => {
    if (response.statusCode == 302) {
      request(response.headers.location, file);
    } else {
      response.pipe(file);
      response.on("end", () => {
        console.log("Downloaded keycloak");
        decompressKeycloak().then(() => run());
      });
    }
  });
}

if (!fs.existsSync(fileName)) {
  const file = fs.createWriteStream(fileName);
  request(`https://github.com/keycloak/keycloak/releases/download/${version}/keycloak-${version}.tar.gz`, file);
```

ssilvert replied:
@edewit Can you make this into a commit suggestion?  Then I'll commit it and we can give it a try.

edewit commented on start.js:
```diff
@@ -1,53 +1,55 @@
-const folder = 'server';
-const fileName = path.join(folder, 'keycloak.tar.gz');
+const folder = "server";
+const fileName = path.join(folder, "keycloak.tar.gz");
```
Let's change this line so that we always download the right version and don't skip the download step because the old file is there
```suggestion
const fileName = path.join(folder, `keycloak-${version}.tar.gz`);
```

Aboullos commented on tests/cypress/integration/masthead_test.spec.js:
```diff
@@ -0,0 +1,70 @@
+  });
+
+  it("shows kabab and hides regular menu", () => {
+    cy.get("#user-dropdown").should("not.exist");
+    cy.get("#user-dropdown-kebab").should("exist");
```
This two can be added to `HeaderPage` something as `checkUserDropdown`
And it could check `#user-dropdown` is present and #user-dropdown-kebab is not when `Cypress.config().viewportWidth < 768` and viceversa

Aboullos replied:
> This two can be added to `Masthead` something as `checkUserDropdown`
> And it could check `#user-dropdown` is present and `#user-dropdown-kebab` is not using `isMobileMode()`

Aboullos replied:
This could be changed to use the elements from `Masthead` to remove the hardcoded strings and in case the element changes in a future
```suggestion
    cy.get(masthead.userDrpDwn).should("not.exist");
    cy.get(masthead.userDrpDwnKebab).should("exist");
```

Aboullos commented on tests/cypress/integration/masthead_test.spec.js:
```diff
@@ -0,0 +1,70 @@
+
+const logOutTest = (userDropdown) => {
+  it("logs out", () => {
+    cy.get(userDropdown).click();
+    cy.get("#sign-out").click();
```
```suggestion
    headerpage.signOut();
```

And modify the signOut() method to user a different dropdown according to the viewport wide 
e.g. `Cypress.config().viewportWidth < 768`

Aboullos commented on tests/cypress/integration/masthead_test.spec.js:
```diff
@@ -0,0 +1,70 @@
+    sidebarPage.goToClientScopes();
+    listingPage.goToItemDetails("address");
+
+    cy.get("#view-header-subkey").should("exist");
+    cy.get(`#${CSS.escape("client-scopes-help:name")}`).should("exist");
+
+    cy.get("#help").click();
+    cy.get("#enableHelp").click({ force: true });
+
+    cy.get("#view-header-subkey").should("not.exist");
+    cy.get(`#${CSS.escape("client-scopes-help:name")}`).should("not.exist");
```
This can be added as two methods in `HeaderPage`, something like `helpIsShown(boolean)` and `switchShowHelp()`

ssilvert commented:
@Aboullos Please re-review.

I have refactored some of this stuff and taken some of your suggestions.  We now have a Masthead class.  I did remove code that was not being used and I moved the dialog code into a utility class because it doesn't really belong as part of the Masthead.

Also note that I have added `id` attributes to the code to make it easier to query parts of the DOM.  We should make this a best practice because `id` attributes a much less brittle than relying on more complex queries.

For various reasons, I did not think it appropriate to make a class for the account console.   We must never navigate to account console in any admin test outside of the tests I wrote here.

Aboullos commented on tests/cypress/support/pages/admin_console/Masthead.js:
```diff
@@ -0,0 +1,62 @@
+
+  toggleGlobalHelp() {
+    cy.get("#help").click();
+    cy.get("#enableHelp").click({ force: true });
```
This would be better to add it as an object in the constructor with the others, to separate elements from actions.
Something like `enableHelpSwitch`

ssilvert replied:
No reason to do this until someone needs it.  See YAGNI comments below.

Aboullos commented on tests/cypress/support/pages/admin_console/Masthead.js:
```diff
@@ -0,0 +1,62 @@
+
+  signOut() {
+    this.userDropdown().click();
+    cy.get("#sign-out").click();
```
This would be better to add it as an object in the constructor with the others, to separate elements from actions.
Something like `signOutBtn` as it was in the previous `HeaderPage`

Aboullos commented on tests/cypress/support/pages/admin_console/Masthead.js:
```diff
@@ -0,0 +1,62 @@
+
+  accountManagement() {
+    this.userDropdown().click();
+    cy.get("#manage-account").click();
```
This would be better to add it as an object in the constructor with the others, to separate elements from actions.
Something like `manageAccountBtn` as it was in the previous `HeaderPage`

Aboullos commented on tests/cypress/integration/masthead_test.spec.js:
```diff
@@ -0,0 +1,69 @@
+    cy.get(`#${CSS.escape("client-scopes-help:name")}`).should("not.exist");
+  });
+
+  logOutTest("#user-dropdown");
```
```suggestion
  logOutTest();
```

Aboullos commented on tests/cypress/integration/masthead_test.spec.js:
```diff
@@ -0,0 +1,69 @@
+  });
+};
+
+const goToAcctMgtTest = (userDropdown) => {
```
```suggestion
const goToAcctMgtTest = () => {
```

ssilvert replied:
Already done

Aboullos commented on tests/cypress/integration/masthead_test.spec.js:
```diff
@@ -0,0 +1,69 @@
+    loginPage.logIn();
+  });
+
+  goToAcctMgtTest("#user-dropdown");
```
```suggestion
  goToAcctMgtTest();
```

ssilvert replied:
already changed that one

Aboullos commented on tests/cypress/integration/masthead_test.spec.js:
```diff
@@ -0,0 +1,69 @@
+  // TODO: Add test for help when using kebab menu.
+  //       Feature not yet implemented for kebab.
+
+  goToAcctMgtTest("#user-dropdown-kebab");
```
```suggestion
  goToAcctMgtTest();
```

ssilvert replied:
already changed that one

Aboullos commented on tests/cypress/integration/masthead_test.spec.js:
```diff
@@ -0,0 +1,69 @@
+    sidebarPage.goToClientScopes();
+    listingPage.goToItemDetails("address");
+
+    cy.get("#view-header-subkey").should("exist");
+    cy.get(`#${CSS.escape("client-scopes-help:name")}`).should("exist");
+
+    masthead.toggleGlobalHelp();
+
+    cy.get("#view-header-subkey").should("not.exist");
+    cy.get(`#${CSS.escape("client-scopes-help:name")}`).should("not.exist");
```
To have this as a reusable method (it could be used in tests to check that all the windows that contain help can show and hide it. It could be in `Masthead` as `helpIsShown(boolean)` where the "exist" or ''not.exist" change according to the boolean value
```suggestion
   masthead.helpIsShown(true);

    masthead.toggleGlobalHelp();

    masthead.helpIsShown(false);
```

ssilvert replied:
We should never hide the assertions in a utility class like Masthead.  You should always be able to read the test and know what it is asserting.  If I put this stuff into Masthead it would hide the entire test.

Also, we don't want to refactor something just because we think we might need it later.  It's extra complexity for no benefit.  Follow the [YAGNI principle](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

ssilvert commented on tests/cypress/support/pages/admin_console/Masthead.js:
```diff
@@ -0,0 +1,62 @@
+export default class Masthead {
+  constructor() {
+    this.isMobileMode = false;
```
fixed
